common/persistence/visibility_hybrid_manager.go

```diff
@@ -0,0 +1,554 @@
+	"github.com/uber/cadence/common/log/tag"
+	"github.com/uber/cadence/common/types"
+)
+
+type (
+	visibilityHybridManager struct {
+		logger                    log.Logger
+		visibilityMgrs            map[string]VisibilityManager
+		readVisibilityStoreName   dynamicconfig.StringPropertyFnWithDomainFilter
+		writeVisibilityStoreName  dynamicconfig.StringPropertyFn
+		logCustomerQueryParameter dynamicconfig.BoolPropertyFnWithDomainFilter
+	}
+)
+
+const (
+	ContextKey           = ResponseComparatorContextKey("visibility-override")
+	dbVisStoreName       = "db"
+	advancedWriteModeOff = "off"
+)
+
+// ResponseComparatorContextKey is for Pinot/ES response comparator. This struct will be passed into ctx as a key.
+type ResponseComparatorContextKey string
+
+type OperationType string
+
+var Operation = struct {
+	LIST  OperationType
+	COUNT OperationType
+}{
+	LIST:  "list",
+	COUNT: "count",
+}
+
+var _ VisibilityManager = (*visibilityHybridManager)(nil)
+
+// NewVisibilityTripleManager create a visibility manager that operate on DB or advanced visibility based on dynamic config.
+// For Pinot migration, Pinot is the destination visibility manager, ES is the source visibility manager, and DB is the fallback.
+// For OpenSearch migration, OS is the destination visibility manager, ES is the source visibility manager, and DB is the fallback.
+func NewVisibilityHybridManager(
+	visibilityMgrs map[string]VisibilityManager,
+	readVisibilityStoreName dynamicconfig.StringPropertyFnWithDomainFilter,
+	writeVisibilityStoreName dynamicconfig.StringPropertyFn,
+	logCustomerQueryParameter dynamicconfig.BoolPropertyFnWithDomainFilter,
+	logger log.Logger,
+) VisibilityManager {
+	if len(visibilityMgrs) == 0 {
+		logger.Fatal("require one of dbVisibilityManager or advancedVisibilityManager")
+		return nil
+	}
+	if logCustomerQueryParameter == nil {
+		logCustomerQueryParameter = dynamicconfig.GetBoolPropertyFnFilteredByDomain(false)
+	}
+	return &visibilityHybridManager{
+		visibilityMgrs:            visibilityMgrs,
+		readVisibilityStoreName:   readVisibilityStoreName,
+		writeVisibilityStoreName:  writeVisibilityStoreName,
+		logger:                    logger,
+		logCustomerQueryParameter: logCustomerQueryParameter,
+	}
+}
+
+func (v *visibilityHybridManager) Close() {
+	for _, mgr := range v.visibilityMgrs {
+		if mgr != nil {
+			mgr.Close()
+		}
+	}
+}
+
+func (v *visibilityHybridManager) GetName() string {
+	storeNames := strings.Split(v.writeVisibilityStoreName(), ",")
+	if v.visibilityMgrs[storeNames[0]] != nil {
+		return v.visibilityMgrs[storeNames[0]].GetName()
+	}
+	return storeNames[0] // return the primary store name
+}
+
+func (v *visibilityHybridManager) RecordWorkflowExecutionStarted(
+	ctx context.Context,
+	request *RecordWorkflowExecutionStartedRequest,
+) error {
+	return v.chooseVisibilityManagerForWrite(
+		ctx,
+		func(storeName string) error {
+			return v.visibilityMgrs[storeName].RecordWorkflowExecutionStarted(ctx, request)
```

Comment:
let's check if `visibilityMgrs` contains `storeName` entry in all these Record* func callbacks.
e.g.
```
mgr, ok := v.visibilityMgrs[storeName]
if !ok {
  return fmt.Errorf("Visibility store manager with name %s not found", storeName)
}
return mgr.RecordWorkflowExecutionStarted(ctx, request)
```

Reply:
In side chooseVisibilityManagerForWrite(), we have already checked if mgr exists before call the callback function
if mgr, ok := v.visibilityMgrs[mode]; ok && mgr != nil {
			if err := visFunc(mode); err != nil {
				errors = append(errors, err.Error())
			}

Reply:
yeah I saw that but still nice to check map existence whenever it's accessed to avoid future misuse causing nil pointer exception